tcljava/doc/TclIndex.3
'\"
'\" Copyright (c) 1996-1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: TclIndex.3,v 1.1 1998/10/14 21:09:08 cvsadmin Exp $
'\" 
.so man.macros
.TH TclIndex 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.TclIndex -- This class provides an efficient way to match a
string with a predefined array of strings.

.SH EXTENDS
tcl.lang.InternalRep

.SH METHODS
.sp
static int \fBget\fR(Interp \fIinterp\fR, TclObject \fItobj\fR, String \fItable[]\fR, String \fImsg\fR, int \fIflags\fR) throws TclException

.SH ARGUMENTS
.AS TclObject "table[]" .
.AP Interp interp
If an error occurs while converting \fItobj\fR to a string object,
an error message is left in the interpreter's result object
unless \fIinterp\fR is NULL.
.AP TclObject tobj
TclObject containing the String to compare to those of \fItable[]\fR.
If the internal representation of \fItobj\fR is not TclString, the
method attempts to convert it to TclString.
.AP String table[]
Table of strings to compare against the string internal to \fItobj\fR.
.AP String msg
If an error occurs, \fImsg\fR is included in the error message to
indicate what was being looked up.
.AP int flags
If TCL.EXACT flag is specified, only exact matches are accepted.
Otherwise, matches include unique abbreviations.
.BE

.SH DESCRIPTION
The \fBTclIndex\fR class provides an efficient way to look up
keywords, switch names, option names, and similar things where the
value of an object must be one of a predefined set of values.  We
commonly use this class to identify the subcommand or switch of a Tcl
command.
.TP
\fBget\fR
The \fBget\fR method compares the internal string value of \fItobj\fR
against each of the strings in \fItable\fR to find a match.  A match
occurs if \fItobj\fR's string value is identical to one of the strings
in \fItable\fR, or if it is a unique abbreviation for exactly one of
the strings in \fItable\fR and the TCL.EXACT \fIflag\fR was not
specified.
.sp
If a match is made, the index of the matching entry is returned, and
the internal representation of \fItobj\fR is modified to hold the
address of \fItable\fR and the index of the matching entry.  If
TclIndex.get() is invoked again with the same tobj and table arguments
(e.g. during a re-invocation of a Tcl command), it returns the
matching index immediately without having to redo the lookup
operation.
.sp
If there is no matching entry, a TclException is thrown, and an error
message is left in \fIinterp\fR's result if \fIinterp\fR isn't NULL.
The \fImsg\fR argument is included in the error message to indicate
what kind of data that was being looked up.

.SH EXAMPLE
.RS
Suppose the \fBfoo\fR command requires the use of one of
the options in the following array of strings:
.CS
static final private String opts[] = {
	"first",
	"second",
	"third"
};
.CE
When a call is made to \fBfoo\fR, we use the \fBget\fR method to find
out which option was specified in the first argument to the \fBfoo\fR
command,\fIargv[1]\fR:
.CS
int opt = TclIndex.get(interp, argv[1], opts, "option", 0);
.CE
If \fIargv[1]\fR's internal string representation is either "second",
"sec", or "s" a match is made.  However, if \fIargv[1]\fR's internal
string representation is "scd", no match is made, and the \fBget\fR
method generates the following error message:
.CS
bad option "scd": must be first, second, or third.
.CE
.RE

.SH "EQUIVALENT C FUNCTIONS"
Tcl_GetIndexFromObj

.SH "SEE ALSO"
InternalRep, TclObject, TclString, TclInteger, createCmd, TclException

.SH KEYWORDS
index, internal representation, object, object type, string, exception


tcljava/ChangeLog
2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * README.jacl:
        * README.tclblend: Mention that Jacl and Tcl Blend
        are production ready and include URLs for build
        notes on the wiki.

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        Update version numbers to 1.3.3.

        * configure:
        * configure.in:
        * src/pkgIndex.tcl:
        * src/jacl/tcl/lang/Interp.java:
        * src/tcljava/tcl/lang/BlendExtension.java:
        * unix/tools/vutil/current:
        * win/makefile.vc:

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * configure: Regen.
        * configure.in: Set default --prefix to /usr/local.

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/tclblend/tcl/lang/Interp.java (evalFile):
        Add TclBlend implementation of Interp.evalFile().

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/jacl/tcl/lang/ExecCmd.java (cmdProc):
        Reimplement JDK 1.3 and newer exec so that two
        threads are created. One reads stdout of the
        subprocess while the other reads stderr. This
        is needed to avoid deadlocking in waitFor()
        when a process writes a lot of stdout and stderr.
        * src/tests/tcljava/tests/exec/ExecLotsOutput.java:
        * tests/inprogress/ExecCmd.test: Add test case
        for the waitFor() exec deadlocking bug.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-branch branch and
 	and tcljava-132-branchpoint tags.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-2005-08-08 tag to mark tree
	state for the 1.3.2 release.

2005-07-28  Mo DeJong  <mdejong@users.sourceforge.net>

	* announce: Update.
	* docs/website/index.html: Update.

2005-07-27  Mo DeJong  <mdejong@users.sourceforge.net>

	* README.jacl: Update msys description and
	mention jaclsh.bat script.
	* README.tclblend: Mention Tcl 8.4.11 and
	Thread 2.6.2 reqs. Add msys url and
	mention jtclsh.bat script.
	* docs/website/faq.html: Update FAQ.
	* docs/website/index.html: Update homepage
	info with new 1.3.2 release.
	* docs/website/mail.html: Add mail page that
	explains how the mailing lists work.

2005-07-23  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Channel.java (initOutput):
	When initializing a channel of type "file",
	set a flag to indicate that the channel
	should be sync()'ed in addition to flushing.
	* src/jacl/tcl/lang/FlushCmd.java: Print
	IO exception stack trace, this should never
	happen but if it did sa stack trace would
	be helpful.
	* src/jacl/tcl/lang/TclOutputStream.java (setSync,
	flushChannel):
	Add flag setting method so that an output
	stream can be marked so that sync() will
	be invoked when flushing.
	* tests/tcl/SocketCmd.test: Add simple socket
	test case that writes to and reads from a socket.

2005-07-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/BinaryCmd.java (FormatNumber,
	ScanNumber):
	Handle edge cases when casting from a double
	to a float. Rewrite scanning code so that
	64 bit long value is used for shifting operation.
	* src/jacl/tcl/lang/InfoCmd.java:
	* src/jacl/tcl/lang/Shell.java:
	* src/jacl/tcl/lang/SubstCmd.java:
	Invoke static method via class name.
	* src/jacl/tcl/lang/Util.java (strtod): Add
	support for "NaN", "Inf", and "-Inf" strings
 	as valid double values.
	* tests/tcl/binary.test: Enable test cases
	that were not being run because they had
	the nonPortable constraint. Java supports
	only big endian primitive types so the
	binary format and scan operations can
	just assume big endian layout. Test that
	assume little endian are not run via
	addition of the littleEndian constraint.
	Add test cases for double -> float edge cases.

2005-07-20  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Copy new files into dist.
	* tests/tclblend/loading.test: Rework test
	so that it is only run under Win32 and
	returns the proper result.
	* win/jaclsh.bat: Removed old bat script.

2005-07-19  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Remove old green threads code
	for "custom" target. Tcl Blend only works
	with a native threads JVM anyway. Add new
	custom2 shell which loads a JVM at runtime
	and then loads Tcl Blend like a JVM would.
	This shell make it easier to debug weird
	loading issues in Tcl Blend.
	* src/tests/tclblend/tests/JavaEval.java:
	Make the program exit(0) when an expected
	result matches the program result. Use
	positive exit status.
	* tests/tclblend/loading.test: Use the
	expected compare test in the JavaEval
	class.
	* unix/custom2.c: Add new custom shell impl
	that loads a JVM and then loads Tcl Blend into
	the JVM.

2005-07-18  Mo DeJong  <mdejong@users.sourceforge.net>

	Fix crashes when loading Tcl Blend into an
	existing JVM by invoking Tcl_FindExecutable("java")
	to init Tcl subsystems.

	* src/native/javaCmd.c: Update comment.
	* src/native/javaInterp.c (
	Java_tcl_lang_Interp_create,
	Java_tcl_lang_Interp_init,
	Java_tcl_lang_Interp_initName): Add
	TCLBLEND_DEBUG output statements.
	Add initName() method that will init the
	Tcl executable name when Tcl Blend is loaded
	from Java.
	* src/tclblend/tcl/lang/Interp.java (shlib_load,
	initName):
	Add initName() method mentioned above and
	invoke it after loading the Tcl Blend shared lib.
	* tests/tclblend/loading.test: Update tests
	to account for executable name check. Add
	test for [info nameofexecutable] in a Tcl Blend
	interp loaded into Java.

2005-07-17  Mo DeJong  <mdejong@users.sourceforge.net>

	Add startup .bat files for Jacl and Tcl Blend.
	The bat files are auto generated by configure
	with the correct paths.

	* Makefile.in: Install and test jaclsh.bat
	and jtclsh.bat. Make sure the PATH variable
	is always set with the new directories
	prepended to the PATH. Rework tricky gdb
	invocation under Win32 since the mingw
	version of gdb does not know about mingw
	style unix paths.

tcljava/src/tcljava/tcl/lang/TclIndex.java
/*
 * TclIndex.java
 *
 *	This file implements objects of type "index".  This object type
 *	is used to lookup a keyword in a table of valid values and cache
 *	the index of the matching entry.
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: TclIndex.java,v 1.3 1999/07/28 03:50:56 mo Exp $
 */

package tcl.lang;

public class TclIndex extends InternalRep {

    /**
     * The variable slots for this object.
     */
    private int index;

    /**
     * Table of valid options.
     */

    private String[] table;

    /**
     * Construct a TclIndex representation with the given index & table.
     */
    private TclIndex(int i, String[] tab) {
	index = i;
	table = tab;
    }

    /**
     * Returns a dupilcate of the current object.
     * @param obj the TclObject that contains this internalRep.
     */
    protected InternalRep duplicate() {
	return new TclIndex(index, table);
    }

    /**
     * Called to query the string representation of the Tcl object. This
     * method is called only by TclObject.toString() when
     * TclObject.stringRep is null.
     *
     * @return the string representation of the Tcl object.
     */
    public String toString() {
	return table[index];
    }

    /**
     * Gets the index into the table of the object.  Generate an error
     * it it doesn't occur.  This also converts the object to an index
     * which should catch the lookup for speed improvement.
     *
     * @param interp the interperter or null
     * @param tobj the object to operate on.
     * @paran table the list of commands
     * @paran msg used as part of any error messages
     * @paran flags may be TCL.EXACT.
     */

    public static int get(Interp interp, TclObject tobj, String[] table,
	    String msg, int flags) throws TclException {
	InternalRep rep = tobj.getInternalRep();

	if (rep instanceof TclIndex) {
	    if (((TclIndex) rep).table == table) {
		return ((TclIndex) rep).index;
	    }
	}

	String str = tobj.toString();
	int strLen = str.length();
	int tableLen = table.length;
	int index = -1;
	int numAbbrev = 0;

	checking: {
	    if (strLen > 0) {

		for (int i = 0; i < tableLen; i++) {
		    String option = table[i];

		    if (((flags & TCL.EXACT) == TCL.EXACT) && 
		            (option.length() != strLen)) {
			continue;
		    }
		    if (option.equals(str)) {
			// Found an exact match already. Return it.

			index = i;
			break checking;
		    }
		    if (option.startsWith(str)) {
			numAbbrev++;
			index = i;
		    }
		}
	    }
	    if (numAbbrev != 1) {
		StringBuffer sbuf = new StringBuffer();
		if (numAbbrev > 1) {
		    sbuf.append("ambiguous ");
		} else {
		    sbuf.append("bad ");
		}
		sbuf.append(msg);
		sbuf.append(" \"");
		sbuf.append(str);
		sbuf.append("\"");
		sbuf.append(": must be ");
		sbuf.append(table[0]);
		for (int i = 1; i < tableLen; i++) {
		    if (i == (tableLen - 1)) {
			sbuf.append(", or ");
		    } else {
			sbuf.append(", ");
		    }
		    sbuf.append(table[i]);
		}
		throw new TclException(interp, sbuf.toString());
	    }
	}

	// Create a new index object.

	tobj.setInternalRep(new TclIndex(index, table));
	return index;
    }
}

